### ISSUE TITLE
Best effort connections.

### ISSUE THREAD

**Issue by Soroosh129** [icyphy/lingua-franca#235]
With the recent change to physical actions, and subsequently physical connections (as discussed in #227), we lose the ability to make a best effort connection that can use a "poor man's Ptides" to preserve timestamps in the centralized coordination mode.

In a prior meeting, we discussed the possibility of getting back that capability by enabling the usage of the `tardy` handler in centralized coordination. Using a sender-receiver example, the following federated program describes the construction of a best-effort logical connection:
```
target C {
    coordination: centralized
};
reactor Sender {
    output out:int;
    ...
    // Send output on out
    ...
}
reactor Receiver {
    input in:int;
    reaction(in) {=
        // Body of the reaction
    =} tardy {=
        // Marking connection (1) as best-effort
        // Alternative body in case of a tardy message
    =}
}
federated reactor BestEffortConnection {
    sender = new Sender();
    receiver = new Receiver();

    sender.out -> receiver.in; // Connection (1)
}
```

However, there is still the outstanding issue of having multiple triggering reactions at the receiver, or perhaps even having multiple downstream reactors (see [DistributedCountDecentralizedLateHierarchy](https://github.com/icyphy/lingua-franca/blob/federated-execution/test/C/DistributedCountDecentralizedLateHierarchy.lf)).

The core questions I have right now are:
1- Does the logical connection (e.g., connection (1)) become best-effort only when all triggered reactions at the receiver have a `tardy` handler?
2- If so, what does it mean to have the `tardy` handler on some of the triggering reactions at the receiver and not the others? Should we issue a warning perhaps?

**Comment by cmnrd** [comment]
I think we should find the answer to another question first: How do we handle faults in LF? If we answer that, it probably becomes easier to answer this question here.

I find it problematic to look at fault handlers as an alternative behavior of reactions. Maybe every reactor should be able to declare central fault handlers that listen for certain types of faults (such as tardiness on specific ports or any port). I am thinking of a similar syntax as is used for catching exceptions in Java and other languages. I am not sure how this could work precisely. But my point is, if we decide to use such a fault handling mechanism, the question above becomes simpler. We do not need to consider all reactions, but only the involved reactors. I really think we should give priority to finding a general fault handling mechanism.

**Comment by edwardalee** [comment]
I don't think we have actually lost the ability to make a best-effort, poor-person's Ptides connection.  Here's how to do it: 

First, make a physical connection with no "after". Second, include in your payload the time-stamp *t* at the sender.  At the receiving end, the message is assigned a new timestamp *T* based on the physical clock. The reaction to this message should schedule a logical action at *t* + *a*, where *a* is the intended "after".  It does this by calling `schedule(action, t + a - T)`. If *t* + *a* < *T*, then this extra delay is negative. The message is tardy.  So the receiving reaction should handle this tardiness in an application-specific way.

I think I would prefer that this be mechanism rather than adding complexity to the language.

**Comment by Soroosh129** [comment]
Yes I think this is a great alternative. However, the caveat is that we lose the microstep information.

Moreover, we need to manually create a separate action for each incoming port that is intended to use Ptides and a duplicate reaction to that action. This is essentially what we were doing before but it was hidden from the programmer.

**Comment by edwardalee** [comment]
You could carry the microstep info in the payload as well and manually implement the skipping reaction.
